Source/WebCore/platform/graphics/cg/NativeImageCG.cpp

```diff
@@ -31,9 +31,35 @@
 #include "CGSubimageCacheWithTimer.h"
 #include "GeometryUtilities.h"
 #include "GraphicsContextCG.h"
+#include <limits>
+#include <pal/spi/cg/CoreGraphicsSPI.h>
 
 namespace WebCore {
 
+RefPtr<NativeImage> NativeImage::create(PlatformImagePtr&& image, RenderingResourceIdentifier renderingResourceIdentifier)
+{
+    if (!image)
+        return nullptr;
+    if (CGImageGetWidth(image.get()) > std::numeric_limits<int>::max() || CGImageGetHeight(image.get()) > std::numeric_limits<int>::max())
+        return nullptr;
+    return adoptRef(*new NativeImage(WTFMove(image), renderingResourceIdentifier));
+}
+
+RefPtr<NativeImage> NativeImage::createTransient(PlatformImagePtr&& image, RenderingResourceIdentifier identifier)
+{
+    if (!image)
+        return nullptr;
+    // FIXME: GraphicsContextCG caching should be made better and this should be the default mode
+    // for NativeImage, as we cannot guarantee all the places that draw images to not cache unwanted
+    // images.
+    RetainPtr<CGImage> transientImage = CGImageCreateCopy(image.get()); // Make a shallow copy so the metadata change doesn't affect the caller.
+    if (!transientImage)
+        return nullptr;
+    image = nullptr;
+    CGImageSetCachingFlags(transientImage.get(), kCGImageCachingTransient);
```

Comment:
I want to make sure I have this straight. When we read from the CGImageRef contained in the `image` variable here (later on, somewhere else) we cause CG to put the underlying data into a global cache? And even though we eventually clear it from our caches the retain count is non-zero because of that global cache?

This SPI then tells CG not to cache the underlying data?

Reply:
When a CGImage is drawn, the underlying storage is sometimes converted to other formats and cropped and the result is stored to global cache. This is done so that the next time these conversions can be skipped.
Transient flag tells CG that likely there won't be next time.
This doesn't fix our underlying problems, it's a small step fixing them.

Reply:
Ah ok! That makes sense.